app/src/main/java/com/woefe/shoppinglist/activity/RecyclerListAdapter.java

```diff
@@ -190,19 +232,28 @@ public boolean isLongPressDragEnabled() {
         }
 
         @Override
-        public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
+        public int getMovementFlags(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
             final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
             final int swipeFlags = ItemTouchHelper.START;
             return makeMovementFlags(dragFlags, swipeFlags);
         }
 
         @Override
-        public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
+        public boolean onMove(@NonNull RecyclerView recyclerView,
+                              RecyclerView.ViewHolder viewHolder,
+                              RecyclerView.ViewHolder target) {
             if (viewHolder.getItemViewType() != target.getItemViewType()) {
                 return false;
             }
 
-            RecyclerListAdapter.this.move(viewHolder.getAdapterPosition(), target.getAdapterPosition());
+            if (!(viewHolder instanceof ItemViewHolder) || !(target instanceof ItemViewHolder)) {
+                return false;
+            }
+
+            ItemViewHolder sourceViewHolder = (ItemViewHolder) viewHolder;
+            ItemViewHolder targetViewHolder = (ItemViewHolder) target;
+
+            RecyclerListAdapter.this.move(sourceViewHolder.listItem, targetViewHolder.listItem);
```

Comment:
Not entirely sure where it was introduced, but when moving items, it only swaps with the neighbor item.

Reply:
I will have a look :+1: